xline/src/server/watch_server.rs

```diff
@@ -0,0 +1,263 @@
+use std::{collections::HashMap, sync::Arc};
+
+use clippy_utilities::OverflowArithmetic;
+use log::debug;
+use tokio::sync::mpsc;
+use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+
+use super::command::KeyRange;
+use crate::{
+    rpc::{RequestUnion, ResponseHeader, Watch, WatchRequest, WatchResponse},
+    storage::kvwatcher::{KvWatcher, WatchEvent, WatchId, Watcher},
+};
+
+/// Watch Server
+#[derive(Debug)]
+pub(crate) struct WatchServer {
+    /// KV watcher
+    watcher: Arc<KvWatcher>,
+}
+
+impl WatchServer {
+    /// New `WatchServer`
+    pub(crate) fn new(watcher: Arc<KvWatcher>) -> Self {
+        Self { watcher }
+    }
+}
+
+/// Handler for one watch connection
+#[derive(Debug)]
+struct WatchHandle {
+    /// KV watcher
+    kv_watcher: Arc<KvWatcher>,
+    /// `WatchResponse` Sender
+    response_tx: mpsc::Sender<Result<WatchResponse, tonic::Status>>,
+    /// `WatchRequest` receiver
+    request_rx: tonic::Streaming<WatchRequest>,
+    /// Event receiver
+    event_rx: mpsc::Receiver<WatchEvent>,
+    /// Event sender
+    event_tx: mpsc::Sender<WatchEvent>,
+    /// Watch ID to watcher map
+    watcher_map: HashMap<WatchId, Watcher>,
+    /// Next available `WatchId`
+    next_id: WatchId,
+}
+
+impl WatchHandle {
+    /// New `WatchHandle`
+    fn new(
+        kv_watcher: Arc<KvWatcher>,
+        response_tx: mpsc::Sender<Result<WatchResponse, tonic::Status>>,
+        request_rx: tonic::Streaming<WatchRequest>,
+        event_rx: mpsc::Receiver<WatchEvent>,
+        event_tx: mpsc::Sender<WatchEvent>,
+    ) -> Self {
+        Self {
+            kv_watcher,
+            response_tx,
+            request_rx,
+            event_rx,
+            event_tx,
+            watcher_map: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    /// Get next available `WatchId`, return -1 if id is not available
+    fn get_next_watch_id(&mut self, id: WatchId) -> WatchId {
+        // 0 means auto-generate
+        if id == 0 {
+            loop {
+                let next = self.next_id;
+                self.next_id = self.next_id.overflow_add(1);
```

Comment:
will we reuse the watch id? what if wrapping happens?

Reply:
each client will have a dedicated watch id pool and watch id is i64. Don't need to reuse.